### PR TITLE
[WIP] Changes all RPC-related database transactions to read-only

### PR DESCRIPTION
**Pull request by tjayrush**
Fixes issue #1279

### CODE REVIEW

**Comment by AskAlexSharov** [comment]
but it's 2 different objects: TxDb and Tx. with different functinality. 
For example TxDb doesn't create new cursors all the time. 
So, i don't think it's good for performance to move high-level code from TxDb to Tx.

**Comment by tjayrush** [comment]
Sorry about that. I just blindly made the change.

This brings up something I've had in the back of my mind. I've been wanting to try to surround the RPC with some performance testing. (Maybe it already is.)

For example, it's very easy for me to run TrueBlocks against the RPC like 5,000 times as quickly as possible for any (or all) of the endpoints.

Before we merge this, I can do that against both `master` and this branch and report back here to help us better make a decision.

**Comment by AskAlexSharov** [comment]
better change api of TxDb

**Comment by AlexeyAkhunov** [comment]
> but it's 2 different objects: TxDb and Tx. with different functinality.
> For example TxDb doesn't create new cursors all the time.
I did not know about this :)

**Comment by AlexeyAkhunov** [comment]
I think it is ok to merge this, because it won't affect performance of RPC daemon, but we can work through re-introducing TxDb for the shared-memory access, which should already be enabled by this change

**Comment by AskAlexSharov** [comment]
This `tx.(ethdb.HasTx).Tx()` probably not work, because `ethdb.Tx` doesn't match this interface.

**Comment by AlexeyAkhunov** [comment]
> This `tx.(ethdb.HasTx).Tx()` probably not work

I think we are not really using that cast anymore in RPC daemon, but @tjayrush can check

**Comment by AlexeyAkhunov** [comment]
I tried this branch. And it is still locked up in the opening of the db. Perhaps we need to add "readonly" option when we are opening the DB:
```
goroutine 1 [syscall, 4 minutes, locked to thread]:
runtime.cgocall(0xe167c0, 0xc000535360, 0xc00029e698)
        runtime/cgocall.go:133 +0x5b fp=0xc000535330 sp=0xc0005352f8 pc=0x42b65b
github.com/ledgerwatch/lmdb-go/lmdb._Cfunc_mdb_txn_begin(0x2027430, 0x0, 0x0, 0xc00029e698, 0x0)
        _cgo_gotypes.go:961 +0x4d fp=0xc000535360 sp=0xc000535330 pc=0x808e8d
github.com/ledgerwatch/lmdb-go/lmdb.beginTxn.func1(0xc0000a8cf0, 0x0, 0x0, 0xc00029e680, 0xc0005353f0)
        github.com/ledgerwatch/lmdb-go@v1.16.0/lmdb/txn.go:99 +0xda fp=0xc0005353a8 sp=0xc000535360 pc=0x81283a
github.com/ledgerwatch/lmdb-go/lmdb.beginTxn(0xc0000a8cf0, 0x0, 0x0, 0x4, 0x1800, 0x2290000010c7180)
        github.com/ledgerwatch/lmdb-go@v1.16.0/lmdb/txn.go:99 +0xb7 fp=0xc0005353f8 sp=0xc0005353a8 pc=0x80e0f7
github.com/ledgerwatch/lmdb-go/lmdb.(*Env).BeginTxn(0xc0000a8cf0, 0x0, 0x0, 0xc0003f8b28, 0xc0003f8cc0, 0xc0003f9610)
        github.com/ledgerwatch/lmdb-go@v1.16.0/lmdb/env.go:435 +0x45 fp=0xc000535450 sp=0xc0005353f8 pc=0x80d585
github.com/ledgerwatch/turbo-geth/ethdb.(*LmdbKV).Begin(0xc00007ca20, 0x13110a0, 0xc00003a080, 0x0, 0x0, 0x43c101, 0xc000535548, 0xae9f42af, 0xc09a8cc18fd033f8, 0xffffffffffffffff)
        github.com/ledgerwatch/turbo-geth/ethdb/kv_lmdb.go:294 +0x8c fp=0xc0005354a0 sp=0xc000535450 pc=0xa094ec
github.com/ledgerwatch/turbo-geth/ethdb.(*LmdbKV).Update(0xc00007ca20, 0x13110a0, 0xc00003a080, 0xc000535618, 0x0, 0x0)
        github.com/ledgerwatch/turbo-geth/ethdb/kv_lmdb.go:400 +0xca fp=0xc000535538 sp=0xc0005354a0 pc=0xa0a22a
github.com/ledgerwatch/turbo-geth/ethdb.LmdbOpts.Open(0x0, 0x7ffc2bf7e6b8, 0x1e, 0x1200958, 0x20000000000, 0x3e8, 0x5, 0xffffffffffffffff, 0x39, 0x8)
        github.com/ledgerwatch/turbo-geth/ethdb/kv_lmdb.go:164 +0xc71 fp=0xc000535818 sp=0xc000535538 pc=0xa08c51
github.com/ledgerwatch/turbo-geth/ethdb.Open(0x7ffc2bf7e6b8, 0x1e, 0x20, 0x1020560, 0x0)
        github.com/ledgerwatch/turbo-geth/ethdb/object_db.go:76 +0x1d9 fp=0xc0005358d8 sp=0xc000535818 pc=0xa20c59
github.com/ledgerwatch/turbo-geth/cmd/rpcdaemon/cli.OpenDB(0x118f44b, 0xe, 0x7ffc2bf7e6b8, 0x1e, 0x7ffc2bf7e707, 0xd, 0x0, 0x0, 0x0, 0x0, ...)
        github.com/ledgerwatch/turbo-geth/cmd/rpcdaemon/cli/config.go:79 +0x2f7 fp=0xc000535978 sp=0xc0005358d8 pc=0xdcc577
main.main.func1(0x1b41660, 0xc0000b6300, 0x0, 0x8, 0x0, 0x0)
        github.com/ledgerwatch/turbo-geth/cmd/rpcdaemon/main.go:20 +0xc5 fp=0xc000535d80 sp=0xc000535978 pc=0xe155e5
github.com/spf13/cobra.(*Command).execute(0x1b41660, 0xc0000320a0, 0x8, 0x8, 0x1b41660, 0xc0000320a0)
        github.com/spf13/cobra@v1.0.0/command.go:842 +0x47c fp=0xc000535e50 sp=0xc000535d80 pc=0x7b9c3c
github.com/spf13/cobra.(*Command).ExecuteC(0x1b41660, 0xc00029e600, 0xc000535f78, 0x42e0a5)
        github.com/spf13/cobra@v1.0.0/command.go:950 +0x375 fp=0xc000535f28 sp=0xc000535e50 pc=0x7ba7b5
github.com/spf13/cobra.(*Command).Execute(...)
        github.com/spf13/cobra@v1.0.0/command.go:887
github.com/spf13/cobra.(*Command).ExecuteContext(...)
        github.com/spf13/cobra@v1.0.0/command.go:880
```

**Comment by AlexeyAkhunov** [comment]
Oh, I think it tried to create buckets, that would probably be skipped if we open in read-only mode

**Comment by tjayrush** [comment]
Just some performance information. Not perfect because super small tests cases, but something:

We ran two sets of tests against four different configurations: `TG master`, `TG branch 1279`, `Parity running remotely`, `Parity running locally`.

---

### Test 1

```[bash]
getBlock 0-8000000:25000 >/dev/null
```

Runs 320 tests (four per 100,000 blocks for 8,000,000 blocks) translating into these RPC commands:

| count | RPC command               |
| ----- | ------------------------- |
| 322   | eth_getBlockByNumber      |
| 36109 | eth_getTransactionReceipt |

with these timings:

| Node          | Timing |
| ------------- | ------ |
| TG master     | 259.62 |
| TG 1279       | 255.04 |
| Parity local  | 33.59  |
| Parity remote | 131.38 |

---

### Test 2

```[bash]
getBlock 0-11000000:250000 --uniq_tx
```

Runs 44 tests (four per milling for 11 million) translating into these RPC commands:

| count | RPC command                       |
| ----- | --------------------------------- |
| 1     | eth_blockNumber                   |
| 44    | eth_getBlockByNumber              |
| 6362  | eth_getTransactionReceipt         |
| 2     | eth_getUncleByBlockNumberAndIndex |
| 44    | eth_getUncleCountByBlockNumber    |
| 3374  | trace_transaction                 |
| 2     | web3_clientVersion                |

---

with these timings:

| Node          | Timing  |
| ------------- | ------- |
| TG master     | 1474.83 |
| TG 1279       | 1429.44 |
| Parity local  | 11.06   |
| Parity remote | 39.03   |

**Comment by tjayrush** [comment]
This code (`tx.(ethdb.HasTx).Tx()`)  appears only at this line in the rpcdaemon: https://github.com/ledgerwatch/turbo-geth/blob/master/cmd/rpcdaemon/commands/debug_api.go#L56.

Very luckily, this is in a function called `StorageRangeAt` which implements `debug_s`torageRangeAt` which fails in the Postman tests.

**Comment by tjayrush** [comment]
From Alex on how to proceed:

```
- tx.(ethdb.HasTx).Tx() - just use tx
- add readOnly when opening db in RPCDaemon
- don't call os.MkdirAll in readOnly mode in kv_lmdb:Open 
- test that things working and merge to master
```

File: cmd/rpcdaemon/commands/eth_block.go
Comment (AskAlexSharov) [inline cmd/rpcdaemon/commands/eth_block.go:19]: @AskAlexSharov @tjayrush , I'm still feeling uncomfortable with this change: 
- because it works by accident. For example in next lines `tx` object used as: `ReadBlockByNumber(tx)`. If you go inside `ReadBlockByNumber` you can find `!errors.Is(err, ethdb.ErrKeyNotFound)` - but ethdb.Tx doesn't return this error - and it satisfy `DatabaseReader` by accident - (I faced this problem last week - when passed Tx to place which expected Db - and I promised to change tx API - rename Get to GetOne or even remove the method).
- we can merge this PR if it doesn't break things and solve problem of localDb use of RPCDaemon. But I will make PR now which will allow open Read tx by TxDb/ObjectDb - and will ask you to use it. Then will take a look how hard to change LmdbTx to not satisfy rawdb.DatabaseReader interface.

**Review by AskAlexSharov — APPROVED** [review]

File: cmd/rpcdaemon/commands/eth_block.go
Comment (AskAlexSharov) [reply cmd/rpcdaemon/commands/eth_block.go:19]: @AlexeyAkhunov , @tjayrush here you go guys, sorry for the mess: https://github.com/ledgerwatch/turbo-geth/pull/1287

Comment (AskAlexSharov) [reply cmd/rpcdaemon/commands/eth_block.go:19]: And the one about protection: https://github.com/ledgerwatch/turbo-geth/pull/1288
hope some day we will drop 1 level of abstraction.

**Comment by AlexeyAkhunov** [comment]
So this PR is now superseded, should we just close it?

**Comment by AskAlexSharov** [comment]
better create new one. this point is still valid:
- don't call os.MkdirAll in readOnly mode in kv_lmdb:Open

File: cmd/rpcdaemon/commands/eth_block.go
Comment (tjayrush) [reply cmd/rpcdaemon/commands/eth_block.go:19]: Trying to catch up. Looks the PR was merged, so I guess you fixed it.

**Pull request closed without merging**
